src/merkle/smt/simple/mod.rs

```diff
@@ -0,0 +1,327 @@
+        <Self as SparseMerkleTree<DEPTH>>::insert(self, key, value)
+    }
+
+    /// Inserts a subtree at the specified index. The depth at which the subtree is inserted is
+    /// computed as `DEPTH - SUBTREE_DEPTH`.
+    ///
+    /// Returns the new root.
+    pub fn set_subtree<const SUBTREE_DEPTH: u8>(
+        &mut self,
+        subtree_insertion_index: u64,
+        subtree: SimpleSmt<SUBTREE_DEPTH>,
+    ) -> Result<RpoDigest, MerkleError> {
+        if SUBTREE_DEPTH > DEPTH {
+            return Err(MerkleError::InvalidSubtreeDepth {
+                subtree_depth: SUBTREE_DEPTH,
+                tree_depth: DEPTH,
+            });
+        }
+
+        // Verify that `subtree_insertion_index` is valid.
+        let subtree_root_insertion_depth = DEPTH - SUBTREE_DEPTH;
+        let subtree_root_index =
+            NodeIndex::new(subtree_root_insertion_depth, subtree_insertion_index)?;
+
+        // add leaves
+        // --------------
+
+        // The subtree's leaf indices live in their own context - i.e. a subtree of depth `d`. If we
+        // insert the subtree at `subtree_insertion_index = 0`, then the subtree leaf indices are
+        // valid as they are. However, consider what happens when we insert at
+        // `subtree_insertion_index = 1`. The first leaf of our subtree now will have index `2^d`;
+        // you can see it as there's a full subtree sitting on its left. In general, for
+        // `subtree_insertion_index = i`, there are `i` subtrees sitting before the subtree we want
+        // to insert, so we need to adjust all its leaves by `i * 2^d`.
+        let leaf_index_shift: u64 = subtree_insertion_index * 2_u64.pow(SUBTREE_DEPTH.into());
+        for (subtree_leaf_idx, leaf_value) in subtree.leaves() {
+            let new_leaf_idx = leaf_index_shift + subtree_leaf_idx;
+            debug_assert!(new_leaf_idx < 2_u64.pow(DEPTH.into()));
+
+            self.leaves.insert(new_leaf_idx, *leaf_value);
+        }
+
+        // add subtree's branch nodes (which includes the root)
+        // --------------
+        for (branch_idx, branch_node) in subtree.inner_nodes {
+            let new_branch_idx = {
+                let new_depth = subtree_root_insertion_depth + branch_idx.depth();
+                let new_value = subtree_insertion_index * 2_u64.pow(branch_idx.depth().into())
+                    + branch_idx.value();
+
+                NodeIndex::new(new_depth, new_value).expect("index guaranteed to be valid")
+            };
+
+            self.inner_nodes.insert(new_branch_idx, branch_node);
+        }
+
+        // recompute nodes starting from subtree root
+        // --------------
+        self.recompute_nodes_from_index_to_root(subtree_root_index, subtree.root);
+
+        Ok(self.root)
+    }
+}
+
+impl<const DEPTH: u8> SparseMerkleTree<DEPTH> for SimpleSmt<DEPTH> {
+    type Key = LeafIndex<DEPTH>;
+    type Value = Word;
+    type Leaf = Word;
+    type Opening<'a> = (MerklePath, Cow<'a, Word>);
```

Comment:
In our codebase we use `RpoDigest` and `Word` interchangeably; I think we should move away from that. Even though they're easy to convert from one to the other, they're semantically different, and we should use the type with the appropriate semantic.

But this is a larger issue; I'll update it (in the trait as well), and create a separate issue

Reply:
As an example of how this can create (minor) issues, the [trait bound on](https://github.com/0xPolygonMiden/crypto/blob/88797c393777c220089378a1b183c006effd98ae/src/merkle/smt/mod.rs#L55) `SparseMerkleTree::Opening` is `From<(MerklePath, Word)>` (when [instantiated for `SimpleSmt`](https://github.com/0xPolygonMiden/crypto/blob/88797c393777c220089378a1b183c006effd98ae/src/merkle/smt/simple/mod.rs#L248)). So I [added](https://github.com/0xPolygonMiden/crypto/blob/88797c393777c220089378a1b183c006effd98ae/src/merkle/path.rs#L184-L188) that conversion to `ValuePath`, even though the natural one to use on that type would be `From<(MerklePath, RpoDigest)>` (given that `ValuePath` stores a `RpoDigest` internally).

Reply:
Yep - let's create an issue for this.

Reply:
https://github.com/0xPolygonMiden/crypto/issues/252